### DISCUSSION TITLE
How to add `DoneActorEvent` types to `guard` definitions in `setup`?

### DISCUSSION THREAD

**Discussion by cmh-echomark** [statelyai/xstate#5129] [Q&A] [answered]
I'm having trouble with upgrading from v4 to v5. One of my problems is that "named" guard methods in the `setup` function cannot be typed without extra boilerplate. As soon as I add `events` into the setup function, guards can no longer infer that they are receiving an "onDone" event from an invoked promise. 

I'm not exactly sure what's considered best-practice for typing such an event in this case.

Here is the full proposed structure: [in TS playground](https://www.typescriptlang.org/play/?ts=5.5.0-dev.20240331#code/JYWwDg9gTgLgBAbzgZwKYwK5gDRwGZQQgAKhIwacAvvmXAEQAeyMAhjKvQNxwD0vcGAE8wqZAC44AVgB0AZnkAoRWkxgAFAkVxBIsZK06dAYwgA7DoxgGarZIkEALYGYDmkllBevq2bUdQAN1QLCURbeyRhUUl6ABEAeRkAJQBRABVkgE1uJ29JMwwQACNUKF9-HRcwDGtwuDsHGGc3AH5xT29qfyo-HVZjGGgJJGQiVABJM0CIAGtUABNSIgpUSQIiZfI0dTshM2M4dQBKAF4APkMjXgAqG7hSdBCFvPsKfHRjFp8NkAazOAAQWIEyc7DgxlYAKg6AwUABMCgGFQcGg+FYABtKDdeJU4DDMPDEBQ4uY1oIkagqFwesdenB-K4MKwoAswlcqsh0s5kKSzOTNEEQjAqMc4BdEHidAS4QChRYZBBajUYDISWSaUYKjpeopRTJjDD2KgALIDFqoTT+UwWVBWcSaODVWrUMUSq1a5r5J1mFUyL1uOAAHyDDAA7o5jcEoABaAOuei0vo+4AwYCY8T0YALDGcZMsY0jPHZ3PiDlGBZiQ3AMBp8yZgDKbFgThRIHNLhRjjKealPpm8zLfaqC1iYxAk2mc0WW1W9Fww5QUGMY-GUwHM7Ic+TWp05j55IA2oudOXd0YmSzR009LEKNz1fz6Nrz7vm650JmFmTE6+jLq-1PE8jHfT96AJKAhF-QCqEXABdPtYK1ACKzJIdz2iNZ6DwFxMWgnUd3xdBIPQ3d6zPIx4iSNJMhyO8c04RC8SQ7pRRpIA)

Abbreviated annotation here:

```js
setup({
  types: {
    context: {}
    events: {} as { type: "DO.RETRY"; thing: number },
    input: {}
    // ...
  },
  actors:{ someInvokedPromise: fromPromise(async ()=>{
    /** Pretend this is fetching from an API that can return true or false */
    return {isDone: true};
  })}, 
  guards: {
    isThisDone: ({event}) => {
      /** 
       * How do I get the `DoneActorEvent` typing in here? 
       * TS complains that "Property 'output' does not exist on type `{ type: "DO.RETRY"}`
       */
      return event.output.isDone;
    },
  },
}).createMachine({
  // ...
  states: {
    idle: {
      description: "Start the machine here",
      invoke: {
        id: "someInvokedPromise", 
        src: "someInvokedPromise",
        onDone: [
          {
            guard: { type: "isThisDone" }, // Do I pass a param? Do I have to do that for every guard and every similar invoked promise?
            target:"done"
          },
         // ...
        ]
      }
    },
    // ...
  }
});
```

**Comment by davidkpiano** [comment, marked as answer]
I don't recommend trying to infer `DoneActorEvent` here, since you must assume that, unless the `guard` is inline, that the `guard` can be called in _any_ transition.

Instead, use params:

```ts
guards: {
  isThisDone: (_, { isDone }) => {
    return isDone;
  }
},

// ...

onDone: {
  guard: {
    type: 'isDone',
    params: ({ event }) => ({
      isDone: event.output.isDone
    })
  }
}
```

**Reply by cmh-echomark** [reply]
Ok great! Thank you
